File: 2000~2999/2805. 나무 자르기.py
import sys
input = sys.stdin.readline

trees, need = map(int, input().split())
trees = list(map(int, input().split()))
most_height_tree = max(trees)
start = 0
end = most_height_tree
mid = (start + end) // 2
result = 0

while True:
    for i in trees:
        if i <= mid:
            continue
        else:
            result += i - mid

    if start <= end:
        if result < need:
            end = mid - 1
            mid = (start + end) // 2
            result = 0
        elif result >= need:
            start = mid + 1
            mid = (start + end) // 2
            result = 0
    else:
        break
print(mid)

File: 2000~2999/2805. 나무 자르기.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int N, M;
    cin >> N >> M;

    int height[N];

    for (int i = 0; i < N; i++)
        cin >> height[i];

    int start = 0;
    int end = *max_element(height, height + N);
    int mid = (start + end) / 2;

    long long result;

    while (start < end)
    {
        mid = (start + end) / 2;
        result = 0;

        for (int i = 0; i < N; i++)
        {
            if (height[i] - mid > 0)
                result += height[i] - mid;
        }

        if (result >= M)
            start = mid + 1;

        else if (result < M)
            end = mid;
    }

    cout << end - 1;
    return 0;
}